packages/lab/src/slider/internal/utils.ts

```diff
@@ -1,158 +1,131 @@
-import { useMemo } from "react";
-import type { SliderValue } from "../types";
-import type { LabeledMark, SliderMark } from "./SliderRailMarks";
+import type { RefObject } from "react";
+import type { SliderChangeHandler, SliderValue, ThumbIndex } from "../types";
 
-const updateValueItemNotPushable = (
-  oldValue: number[],
-  index: number,
-  valueItem: number,
+export const getValue = (
+  trackRef: RefObject<Element>,
   min: number,
   max: number,
+  step: number,
+  clientX: number,
 ) => {
-  const newValue = [...oldValue];
-  if (valueItem < oldValue[index]) {
-    const constraint = index === 0 ? min : newValue[index - 1];
-    newValue[index] = Math.max(constraint, valueItem);
-  } else {
-    const constraint =
-      index === newValue.length - 1 ? max : newValue[index + 1];
-    newValue[index] = Math.min(constraint, valueItem);
+  const { width, x } = trackRef.current!.getBoundingClientRect();
+  const localX = clientX - x;
+  const normaliseBetweenValues = (localX / width) * (max - min) + min;
+  const roundedToStep = roundToStep(normaliseBetweenValues, step);
+  const decimals = countDecimalPlaces(step);
+  const rounded = Number(roundedToStep.toFixed(decimals));
+  const value = clampValue(rounded, [min, max]);
+  return value;
+};
+
+export const setValue = (
+  value: SliderValue,
+  newValue: number,
+  index: ThumbIndex,
+  onChange: SliderChangeHandler,
+) => {
+  if (value.length === 2) {
+    const newValueArray = [...value];
+    newValueArray.splice(index, 1, newValue);
+    onChange(newValueArray as SliderValue);
+    return;
+  }
```

Comment:
Not necessarily an issue but this is a mix between flexible and inflexible code.

value.length === 2 but then value is spread and spliced instead